jvmhs/src/Jvmhs/Data/Code.hs

```diff
@@ -433,155 +464,217 @@ instance ToJSON (B.CmpOpr) where
+
+nameAndTypeFromFieldId :: B.FieldId -> Text.Text
+nameAndTypeFromFieldId (B.FieldId (B.NameAndType name type_)) =
+  Text.concat [name,":",B.toText type_]
 
 
+getInClassMethod ::B.AbsMethodId B.High -> [Pair]
+getInClassMethod (B.InClass a b) =
+  [ "class" .= B.classNameAsText a
+  , "method" .= methodIDToText b
+  ]
+
+
+getInvocationAttributes :: B.Invocation B.High -> [Pair]
+getInvocationAttributes = \case
+    B.InvkSpecial avmi ->
+      ("kind" .= String "special")
+      : getAbsVariableMethodId avmi
+
+    B.InvkVirtual abs' ->
+      ("kind" .= String "virtual")
+      : getInClassMethod abs'
+
+    B.InvkStatic avmi ->
+      ("kind" .= String "static")
+     : getAbsVariableMethodId avmi
+
+    B.InvkInterface word8 avmi ->
+       ["kind" .= String "interface"
+       , "word8" .= word8
+       ] ++ getAbsInterfaceMethodId avmi
+
+    B.InvkDynamic invokeDynamicMethod ->
+      ("kind" .= String "dynamic")
+     : getInvokeDynamicMethod invokeDynamicMethod
+
+
+getAbsInterfaceMethodId :: B.AbsInterfaceMethodId B.High -> [Pair]
+getAbsInterfaceMethodId (B.AbsInterfaceMethodId interfaceMethodId) 
+    = getInClassMethod interfaceMethodId
+
+
+getAbsVariableMethodId :: B.AbsVariableMethodId B.High -> [Pair]
+getAbsVariableMethodId = \case
+  B.VInterfaceMethodId a ->
+    [ "method" .= a
+    ]
+  B.VMethodId abs' ->
+    getInClassMethod abs'
+
+getInvokeDynamicMethod :: B.InvokeDynamic B.High -> [Pair]
+getInvokeDynamicMethod = \case 
+  (B.InvokeDynamic attrIndex methodid) ->
+    [ "attr_index" .= attrIndex
+    , "method" .= methodIDToText methodid
+    ] 
+
+getArrayType :: B.ExactArrayType B.High -> [Pair]
+getArrayType = \case
+    B.EABoolean -> ["type" .= getSimpleTypeString "boolean"]
+    B.EAByte -> ["type" .= getSimpleTypeString "byte"]
+    B.EAChar -> ["type" .= getSimpleTypeString "char"]
+    B.EAShort -> ["type" .= getSimpleTypeString "short"]
+    B.EAInt -> ["type" .= getSimpleTypeString "integer"]
+    B.EALong -> ["type" .= getSimpleTypeString "long"]
+    B.EAFloat -> ["type" .= getSimpleTypeString "float"]
+    B.EADouble -> ["type" .= getSimpleTypeString "double"]
+    B.EARef ref -> 
+      ["type" .= getSimpleTypeString "ref"
+      , "ref" .= ref
+      ]
+
+getTypeInfo :: B.VerificationTypeInfo B.High -> [Pair]
+getTypeInfo = \case
+  B.VTTop -> ["type" .= getSimpleTypeString "top"]
+  B.VTInteger -> ["type" .= getSimpleTypeString "integer"]
+  B.VTFloat -> ["type" .= getSimpleTypeString "float"]
+  B.VTLong -> ["type" .= getSimpleTypeString "long"]
+  B.VTDouble -> ["type" .= getSimpleTypeString "double"]
+  B.VTNull -> ["type" .= getSimpleTypeString "null"]
+  B.VTObject ref ->
+    [ "type" .= getSimpleTypeString "object"
+    , "ref" .= ref
+    ]
+  B.VTUninitialized word ->
+    [ "type" .= getSimpleTypeString "uninitialized"
+    , "word" .= word
+    ]
+  B.VTUninitializedThis  -> ["type" .= getSimpleTypeString "null"]
+
+
+getSimpleTypeString :: Text.Text -> Text.Text
```

Comment:
I understood the need for a SimpleType. But I didn't understand the purpose of having a ComplexType. 
Why do we want to wrap up the value (like the class name or array inner reference) in the type?

Can't we just have a single type called 'SimpleType', where we only capture the type information?

Reply:
I think there are two levels of type information. Either it is very simple like "I" or it 
is more complex like {"type" : "array", "arraycontent": "java/lang/String"}. 
We can compress everything into a string, but then it would look like:
"[Ljava/lang/String;"

Reply:
So I am assuming even String is of type 'ref', and the classname is "java/lang/String".

Reply:
Yes! :)

Reply:
I completed those changes.